support-frontend/assets/pages/[countryGroupId]/components/retryPaymentStatus.ts

```diff
@@ -0,0 +1,52 @@
+import type { StatusResponse } from 'helpers/forms/paymentIntegrations/readerRevenueApis';
+
+// ----- Setup ----- //
+const DEFAULT_POLLING_INTERVAL_MILLIS = 3000;
+const DEFAULT_MAX_POLLS = 10;
+
+// corrects ES5 instanceOf error for retry mechanism
+// https://dev.to/dguo/how-to-fix-instanceof-not-working-for-custom-errors-in-typescript-4amp
+class StillPendingError extends Error {
+	constructor(message: string, public trackingUri: string) {
+		super(message);
+		Object.setPrototypeOf(this, StillPendingError.prototype);
+		this.trackingUri = trackingUri;
+	}
+}
+
+function timeOut(milliseconds: number | undefined): Promise<void> {
+	return new Promise((resolve) => setTimeout(resolve, milliseconds));
+}
+
+export function retryPaymentStatus(
+	promiseFunction: () => Promise<StatusResponse>,
+	pollMax: number = DEFAULT_MAX_POLLS,
+	pollInterval: number = DEFAULT_POLLING_INTERVAL_MILLIS,
+	onRetry?: (pollCount: number, pollTimeTotal: number) => void,
+): Promise<StatusResponse> {
+	async function retryPollAndPromise(polls: number): Promise<StatusResponse> {
+		try {
+			if (polls > 0) {
+				await timeOut(pollInterval); // retry
+			}
+			const result = await promiseFunction();
+			if (result.status === 'pending') {
+				throw new StillPendingError('status is pending', result.trackingUri); // retry on pending
```

Comment:
Why do we throw here and then run the retry logic in the catch block? Couldn't we move that logic here?

Reply:
It's a good question! I think Paul and I talked about this and left it like this because it avoids duplicating some logic (both places would have to do the comparison of polls and pollMax to know if we were at the limit or not). Maybe that's better though, what do you think?

Reply:
I haven't completely thought through the alternative, but I would generally try to avoid using exceptions to control normal execution flow ie. things which are not actually errors

Reply:
Ok, have re-factored to  to avoid using exceptions to control normal (pending) execution flow. Does this approach  look ok?